### ISSUE TITLE
retrain custom model

### ISSUE THREAD

**Issue by Favi0** [AlexeyAB/darknet#2173]
I have a custom model inspired in tinyYolov2 and pretrained weights of this new model trained in VOC dataset. Now i want to do some transfer learning for my dataset (classes=1) 

now if i want to do inference with those pretrained weights with the command:
**!./darknet detect cfg/yolonew.cfg yolonew.weights image.png** 

i got this error : 

_Error: in the file data/coco.names number of names 80 that isn't equal to classes=1 in the file cfg/yolonew.cfg
data/horses.jpg: Predicted in 0.000000 milli-seconds._

now if i want to train a model with my dataset, i get after running this command: 

**!./darknet detector train cfg/obj.data yolonew.cfg yolonew.weights -dont_show**  

i get: 

_Region Avg IOU: nan, Class: -nan, Obj: nan, No Obj: nan, Avg Recall: 0.000000,  count: 2
Region Avg IOU: nan, Class: -nan, Obj: nan, No Obj: nan, Avg Recall: 0.000000,  count: 2_
and so on..
    

**!./darknet test cfg/obj.data yolonew.cfg yolonew.weights data/horses.jpg**

_terminate called after throwing an instance of 'std::out_of_range'
  what():  basic_string::substr: __pos (which is 140) > this->size() (which is 109)_

after reading the documentation i see that those commands from above are used for yolov3 right?
so my questions are

which command should i try for inference with the pretrained weights so it can reference the VOC labels or my only label?
which command should i try for training with my custom dataset/obj files

**Comment by Deadmin1** [comment]
> I have a custom model inspired in tinyYolov2 and pretrained weights of this new model trained in VOC dataset. Now i want to do some transfer learning for my dataset (classes=1)
> now if i want to do inference with those pretrained weights with the command:
!./darknet detect cfg/yolonew.cfg yolonew.weights image.png
> i got this error :
>Error: in the file data/coco.names number of names 80 that isn't equal to classes=1 in the file cfg/yolonew.cfg
data/horses.jpg: Predicted in 0.000000 milli-seconds.

You forgot to pass your obj.data file. 
!./darknet detect cfg/yolonew.cfg yolonew.weights image.png
should be
!./darknet detector test cfg/obj.data cfg/yolonew.cfg yolonew.weights image.png

your obj.data should be 
classes = 1
and pointing to the right names file with your 1 class name in it. 

Like it says. coco.names has 80 classes in it and you are trying just to detect 1 class (classes=1).
Make a new .names file and just put 1 line with your classname in it and try again.

**Comment by Favi0** [comment]
> > I have a custom model inspired in tinyYolov2 and pretrained weights of this new model trained in VOC dataset. Now i want to do some transfer learning for my dataset (classes=1)
> > now if i want to do inference with those pretrained weights with the command:
> > !./darknet detect cfg/yolonew.cfg yolonew.weights image.png
> > i got this error :
> > Error: in the file data/coco.names number of names 80 that isn't equal to classes=1 in the file cfg/yolonew.cfg
> > data/horses.jpg: Predicted in 0.000000 milli-seconds.
> 
> You forgot to pass your obj.data file.
> !./darknet detect cfg/yolonew.cfg yolonew.weights image.png
> should be
> !./darknet detector test cfg/obj.data cfg/yolonew.cfg yolonew.weights image.png
> 
> your obj.data should be
> classes = 1
> and pointing to the right names file with your 1 class name in it.
> And the rest should be pointing to the right files/folders to.
> 
> Like it says. coco.names has 80 classes in it and you are trying just to detect 1 class (classes=1).
> Make a new .names file and just put 1 line with your classname in it and link to it from the obj.data file.

hello deadmin thanks for the comment

as i pointed early , i already ran that command! (last command) for some reason i get a weird error , but in that case i should try with classes=20 and change the numbers of filter back to the original because the net is not trained yet for my only class

**Comment by Deadmin1** [comment]
> as i pointed early , i already ran that command! (last command) for some reason i get a weird error , but in that case i should try with classes=20 and change the numbers of filter back to the original because the net is not trained yet for my only class

Yeah you are right. Didnt read it right.

As far as i know you have to reset the original state if you have pretrained settings. So if it was trained with the 20 classes it should work when you redo the changes like you said in the obj.data, obj.names and the cfg file with the filters. 

Your trainingerrors looks like it could be from your trainingfiles. Can you post 2-3 lines from the labelfiles?

**Comment by Favi0** [comment]
----obj.data file
classes= 1  
train  = data/obj/train.txt
valid  = data/obj/test.txt
names = obj.names  
backup = backup/

----obj.names file
PATENTE

--- train.txt 
data/obj/1933.png
data/obj/1960.JPG


-- my network cfg file 
[net]
# Testing
batch=64
subdivisions=32
# Training
# batch=32
# subdivisions=1
width=224
height=224
channels=3
momentum=0.9
decay=0.0005
angle=0
saturation = 1.5
exposure = 1.5
hue=.1

learning_rate=0.001
burn_in=1000
max_batches = 600200
policy=steps
steps=400000,450000
scales=.1,.1

[convolutional]
batch_normalize=0
filters=16
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=32
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=64
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

[convolutional]
batch_normalize=0
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2


[convolutional]
batch_normalize=0
filters=128
size=3
stride=1
pad=1
activation=leaky

[maxpool]
size=2
stride=2

###########

[convolutional]
batch_normalize=0
size=3
stride=1
pad=1
filters=256
activation=leaky

[convolutional]
size=1
stride=1
pad=1
filters=30  **#change from 125 to 30?**
activation=linear

[region]
anchors =  1.08,1.19,  3.42,4.41,  6.63,11.38,  9.42,5.11,  16.62,10.52
bias_match=1
classes=1  **#change from 20 to 1?**
coords=4
num=5
softmax=1
jitter=.2
rescore=0

object_scale=5
noobject_scale=1
class_scale=1
coord_scale=1

absolute=1
thresh = .8
random=1

**Comment by Favi0** [comment]
thats all files i got , i dont have a txt with labels just the names file

**Comment by Deadmin1** [comment]
Ok so let me see if i can help you a little bit. 
You are training 1 class and the name is Patente. 
Your 1933.txt looks right. It has to be 0 because its the 0 position in the textfile (if you had 2 classes it would be 0 for the first in .name and 1 for the second class).

But i think the filters is wrong. I also trained 1 class prediction and in every convolution layer before region you have 30. 

Here from the How to train

So if classes=1 then should be filters=18. If classes=2 then write filters=21.
(Do not write in the cfg-file: filters=(classes + 5)x3)

In your case it should be filters=18 because just 1 class to detect.

**Comment by Favi0** [comment]
ok i think now im understanding the idea behind the files

i went back to the files and left 20 for classes and 125 for filter , like the original working network , also modified the obj.names with VOC labels
i got  after runnin
**!./darknet test cfg/obj.data yolonew.cfg yolonew.weights data/horses.jpg**

0 conv     16  3 x 3 / 1   224 x 224 x   3   ->   224 x 224 x  16 0.043 BF
   1 max          2 x 2 / 2   224 x 224 x  16   ->   112 x 112 x  16 0.001 BF
   2 conv     32  3 x 3 / 1   112 x 112 x  16   ->   112 x 112 x  32 0.116 BF
   3 max          2 x 2 / 2   112 x 112 x  32   ->    56 x  56 x  32 0.000 BF
   4 conv     64  3 x 3 / 1    56 x  56 x  32   ->    56 x  56 x  64 0.116 BF
   5 max          2 x 2 / 2    56 x  56 x  64   ->    28 x  28 x  64 0.000 BF
   6 conv    128  3 x 3 / 1    28 x  28 x  64   ->    28 x  28 x 128 0.116 BF
   7 max          2 x 2 / 2    28 x  28 x 128   ->    14 x  14 x 128 0.000 BF
   8 conv    128  3 x 3 / 1    14 x  14 x 128   ->    14 x  14 x 128 0.058 BF
   9 max          2 x 2 / 2    14 x  14 x 128   ->     7 x   7 x 128 0.000 BF
  10 conv    256  3 x 3 / 1     7 x   7 x 128   ->     7 x   7 x 256 0.029 BF
  11 conv    125  1 x 1 / 1     7 x   7 x 256   ->     7 x   7 x 125 0.003 BF
  12 detection
mask_scale: Using default '1.000000'
Total BFLOPS 0.482 
Loading weights from yoloLite.weights...
 seen 64 
Done!
data/horses.jpg: Predicted in 0.000000 milli-seconds.
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
aeroplane: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
tvmonitor: 100%
aeroplane: 100%
aeroplane: 100%

which is BS prediction 

the original cfg file and weights works , i already used them in darkflow (tf wrapper of darknet) so they should work here aswell 

gonna try the training tip you gave me

**Comment by Favi0** [comment]
update: seems like i had a problem with the weights file , gonna check on that

**Comment by Deadmin1** [comment]
Cool. Keep going.

**Comment by limyudd** [comment]
> @Deadmin1
> ok i think now im understanding the idea behind the files
> 
> i went back to the files and left 20 for classes and 125 for filter , like the original working network , also modified the obj.names with VOC labels
> i got after runnin
> **!./darknet test cfg/obj.data yolonew.cfg yolonew.weights data/horses.jpg**
> 
> 0 conv 16 3 x 3 / 1 224 x 224 x 3 -> 224 x 224 x 16 0.043 BF
> 1 max 2 x 2 / 2 224 x 224 x 16 -> 112 x 112 x 16 0.001 BF
> 2 conv 32 3 x 3 / 1 112 x 112 x 16 -> 112 x 112 x 32 0.116 BF
> 3 max 2 x 2 / 2 112 x 112 x 32 -> 56 x 56 x 32 0.000 BF
> 4 conv 64 3 x 3 / 1 56 x 56 x 32 -> 56 x 56 x 64 0.116 BF
> 5 max 2 x 2 / 2 56 x 56 x 64 -> 28 x 28 x 64 0.000 BF
> 6 conv 128 3 x 3 / 1 28 x 28 x 64 -> 28 x 28 x 128 0.116 BF
> 7 max 2 x 2 / 2 28 x 28 x 128 -> 14 x 14 x 128 0.000 BF
> 8 conv 128 3 x 3 / 1 14 x 14 x 128 -> 14 x 14 x 128 0.058 BF
> 9 max 2 x 2 / 2 14 x 14 x 128 -> 7 x 7 x 128 0.000 BF
> 10 conv 256 3 x 3 / 1 7 x 7 x 128 -> 7 x 7 x 256 0.029 BF
> 11 conv 125 1 x 1 / 1 7 x 7 x 256 -> 7 x 7 x 125 0.003 BF
> 12 detection
> mask_scale: Using default '1.000000'
> Total BFLOPS 0.482
> Loading weights from yolonew.weights...
> seen 64
> Done!
> data/horses.jpg: Predicted in 0.000000 milli-seconds.
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> aeroplane: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> tvmonitor: 100%
> aeroplane: 100%
> aeroplane: 100%
> 
> which is BS prediction
> 
> the original cfg file and weights works , i already used them in darkflow (tf wrapper of darknet) so they should work here aswell (the inference part , with the correct command perhaps? @AlexeyAB )
> 
> gonna try the training tip you gave me for my only class

yolo v2 set filter = (classes + 5) * 5
yolo v3-tiny set filter = (classes + 5) * 3
